<!DOCTYPE html>
<html lang="en-us">
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <meta name="generator" content="Hugo 0.114.1">
    <title>StereoSeq Data Workflow</title>

    <!-- Bootstrap CSS -->
    <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/css/bootstrap.min.css" integrity="sha384-MCw98/SFnGE8fJT3GXwEOngsV7Zt27NXFoaoApmYm81iuXoPkFOJwJ8ERdknLPMO" crossorigin="anonymous">
    <!-- Font Awesome for icons -->
    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.0.13/css/all.css" integrity="sha384-DNOHZ68U8hZfKXOrtjWvjxusGo9WQnrNx2sqG0tfsghAvtVlRW3tvkXWZh58N9jp" crossorigin="anonymous">
    <!-- Custom Styles -->
    <link rel="stylesheet" href="/css/sidebarStyle.css">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.5.1/styles/default.min.css">
    <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.5.1/highlight.min.js"></script>
    <script>hljs.highlightAll();</script>


    <style>
        .btn {
            margin-top: 10px;
        }

        .content-section {
            margin-left: 0;
            padding-left: 0;
        }
    </style>
</head>
<body>
    <nav class="navbar navbar-expand-lg navbar-light bg-white">
        <div class="container">
            <a class="navbar-brand" href="#">StereoSeq Data Workflow</a>
            <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarNav" aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation">
                <span class="navbar-toggler-icon"></span>
            </button>
            <div class="collapse navbar-collapse" id="navbarNav">
                <ul class="navbar-nav ml-auto">
                    <li class="nav-item">
                        <a class="nav-link" href="#">Home</a>
                    </li>
                </ul>
            </div>
        </div>
    </nav>

    <div class="container mt-4">
        <div class="row">
            <!-- Sidebar -->
            <div class="col-md-3">
                <div class="sticky-top" style="top: 20px;">
                    <h3>Navigation</h3>
                    <ul class="nav flex-column">
                        <li class="nav-item"><a href="#preparation" class="nav-link">Preparation</a></li>
                        <li class="nav-item"><a href="#loadData" class="nav-link">Load Data</a></li>
                        <li class="nav-item"><a href="#Preprocessing" class="nav-link">Preprocessing</a></li>
                        <li class="nav-item"><a href="#Normalization" class="nav-link">Normalization</a></li>
                        <li class="nav-item"><a href="#Clustering" class="nav-link">Clustering</a></li>
                        <li class="nav-item"><a href="#Expression Enhancement" class="nav-link">Expression Enhancement</a></li>
                        <li class="nav-item"><a href="#Spatially variable genes identification" class="nav-link">Spatially variable genes identification</a></li>
                        <li class="nav-item"><a href="#Spatial Domain Detection" class="nav-link">Spatial Domain Detection</a></li>
                        <li class="nav-item"><a href="#Cell Type Annotation" class="nav-link">Cell Type Annotation</a></li>
                        <li class="nav-item"><a href="#Cell Segmentation" class="nav-link">Cell Segmentation</a></li>
                        <li class="nav-item"><a href="#Cell cell communications" class="nav-link">Cell cell communications</a></li>
                        <h4>==============</h4>
                        <h4>Recommended tools</h4>
                        <li class="nav-item"><a href="#Giotto" class="nav-link">Giotto</a></li>
                        <!-- More navigation items -->
                    </ul>
                </div>
            </div>

            <!-- Main Content -->
            <div class="col-md-9 content-section">
                <h1 class="mb-4">StereoSeq Data Workflow</h1>
                <h2 class="mb-4" style="font-size: 20px">Created by Stomics US Data Team</h2>
                <p>This is a StereoSeq data analysis workflow. This demonstration will use a data set <a href="http://116.6.21.110:8090/share/21bb9df9-e6c5-47c5-9aa8-29f2d23a6df4" target="_blank" rel="noopener noreferrer">SS200000135TL_D1</a> of a mouse brain.</p>
                <div style="text-align: center;">
                    <img src="images/Image_20240927214042.png" alt="Description of image" style="width:300px; height:auto;">
                </div>
                <!-- Preparation Section -->
                <h2 id="preparation" class="mt-5">Preparation</h2>
                <p>In this section, we will go over the preparation steps necessary to run the StereoSeq data analysis. This includes setting up the environment and installing the required tools and dependencies.</p>

                <h3>Environment Setup</h3>
                <p>Ensure that the following dependencies are installed:</p>

                <!-- Code block toggle button -->
                <button class="btn btn-success float-right" type="button" data-toggle="collapse" data-target="#envCode" aria-expanded="false" aria-controls="envCode">
                    Show/Hide Code
                </button>

                <!-- Collapsible code block -->
                <div class="collapse" id="envCode">
                    <pre><code class="language-bash">
# Install necessary packages
conda create -n stereoSeq_env python=3.8
conda activate stereoSeq_env
pip install numpy pandas scanpy
        </code></pre>
                </div>

                <!-- loadData Section -->
                <h2 id="loadData" class="mt-5">Load Data</h2>
                <p>Load data in R from rds file</p>
                <!-- Code block toggle button -->
                <button class="btn btn-success float-right" type="button" data-toggle="collapse" data-target="#loadRcode" aria-expanded="false" aria-controls="loadRcode">
                    Show/Hide Code
                </button>

                <!-- Collapsible code block -->
                <div class="collapse" id="loadRcode">
                    <pre><code class="language-r">
    library(Seurat)
    data_path="./SS200000135TL_D1.tissue.gef_bin20.seurat.rds"
    st_data=readRDS(data_path)
</code></pre>

                    <p>Load data in python from h5ad file</p>
                    <!-- Code block toggle button -->
                    <button class="btn btn-success float-right" type="button" data-toggle="collapse" data-target="#loadPythoncode" aria-expanded="false" aria-controls="loadPythoncode">
                        Show/Hide Code
                    </button>

                    <!-- Collapsible code block -->
                    <div class="collapse" id="loadPythoncode">
                        <pre><code class="language-python">
      import scanpy as sc
      data_path="./SS200000135TL_D1.tissue.gef_bin20.annData.h5ad"
      st_data = sc.read_h5ad(data_path)
</code></pre>

                    </div>


                    <!-- Preprocessing Section -->
                    <h2 id="Preprocessing" class="mt-5">Preprocessing</h2>
                    <h3>Filtering low quality spots</h3>
                    <p>
                        In spatial transcriptomics data, filtering low-quality spots is essential to ensure reliable downstream analyses, and using robust metrics like the Median Absolute Deviation (MAD) helps identify outliers effectively.
                        The MAD method provides a robust statistic of variability, minimizing the influence of extreme values that may arise from technical noise or damaged tissue regions.
                        When applying MAD-based filtering, spots with extreme counts or features that deviate significantly from the median by a set threshold, such as 5 MADs, are flagged and considered for removal.
                        This approach helps eliminate spots with anomalously low gene counts or excessive count depth that may result from issues like broken membranes or sequencing errors.
                    </p>
                    <p>
                        Furthermore, filtering by the fraction of mitochondrial reads is crucial because spots with a high proportion of mitochondrial counts often indicate compromised cell integrity. These compromised spots may have lost most cytoplasmic RNA, leaving only mitochondrial transcripts, which can skew the data.
                        However, it is important to apply this filter carefully, as certain cell types or tissue regions might naturally have higher mitochondrial content due to biological functions related to respiration.
                    </p>
                    <p>
                        By jointly considering MAD-based thresholds for count depth and filtering based on mitochondrial read fractions, researchers can strike a balance between removing poor-quality data and retaining meaningful biological variation.
                        This approach ensures that only genuinely low-quality spots are filtered out, preserving spatial and functional tissue heterogeneity.
                        Ultimately, these filtering steps improve the accuracy and interpretability of spatial transcriptomics analysis by focusing on reliable, biologically relevant data.
                    </p>

                    <!-- Code block toggle button -->
                    <button class="btn btn-success float-right" type="button" data-toggle="collapse" data-target="#preparationCode" aria-expanded="false" aria-controls="preparationCode">
                        Show/Hide Code
                    </button>

                    <!-- Collapsible code block -->
                    <div class="collapse" id="preparationCode">
                        <pre><code class="language-r">
    st_data=readRDS(data_path)

    preProcessing = function(rds_data){
    metadata=rds_data@meta.data
    median_nCount <- median(metadata$nCount_Spatial, na.rm=TRUE)
                     abs_dev_nCount <- abs(metadata$nCount_Spatial - median_nCount)
  mad_nCount <- median(abs_dev_nCount, na.rm = TRUE)

  median_nFeature <- median(metadata$nFeature_Spatial, na.rm = TRUE)
  abs_dev_nFeature <- abs(metadata$nFeature_Spatial - median_nFeature)
  mad_nFeature <- median(abs_dev_nFeature, na.rm = TRUE)

  lower_nCount <- median_nCount - 5 * mad_nCount
  upper_nCount <- median_nCount + 5 * mad_nCount

  lower_nFeature <- median_nFeature - 5 * mad_nFeature
  upper_nFeature <- median_nFeature + 5 * mad_nFeature

  rds_data.filter <- subset(rds_data, subset = nCount_Spatial  > lower_nCount &
                              nCount_Spatial < upper_nCount &
                              nFeature_Spatial  > lower_nFeature &
                              nFeature_Spatial < upper_nFeature &
                              percent.mito < 10)

  return (rds_data.filter)

}

st_data.filter=preProcessing(st_data)

</code></pre>
                    </div>
                    <img src="images/Filtering.png" alt="Description of image" style="width:600px; height:auto;">



                    <!-- Normalization Section -->
                    <h2 id="Normalization" class="mt-5">Normalization</h2>
                    <p>
                        Normalization in spatial transcriptomics data is essential to account for technical variability and ensure that differences in gene expression across spatial locations reflect true biological variation rather than artifacts. Factors like differences in sequencing depth, RNA capture efficiency, and background noise can introduce biases that obscure meaningful patterns. By normalizing the data, we adjust for these inconsistencies, allowing for accurate comparison of gene expression levels across spatial regions and facilitating reliable downstream analyses, such as identifying spatially variable genes and constructing spatial gene expression maps.
                    </p>

                    <h3>LogNormalize</h3>

                    <!-- Code block toggle button -->
                    <button class="btn btn-success float-right" type="button" data-toggle="collapse" data-target="#LogNormalizeCode" aria-expanded="false" aria-controls="LogNormalizeCode">
                        Show/Hide Code
                    </button>
                    <!-- Collapsible code block -->
                    <div class="collapse" id="LogNormalizeCode">
                        <pre><code class="language-r">
    st_data <- NormalizeData(
                     st_data.filter,
                     assay='RNA' ,
                     normalization.method="LogNormalize" )

</code></pre>

                        <img src="images/LogNormalize.png" alt="Description of image" style="width:600px; height:auto;">


                    </div>

                    <!-- Clustering Section -->
                    <h2 id="Clustering" class="mt-5">Clustering</h2>
                    <p>
                        Clustering is a technique used to identify groups of cells with similar gene expression profiles, providing insights into the cellular composition and spatial organization of tissues.
                    </p>

                    <h3>Clustering using seurat</h3>

                    <!-- Code block toggle button -->
                    <button class="btn btn-success float-right" type="button" data-toggle="collapse" data-target="#SeuratClusteringCode" aria-expanded="false" aria-controls="SeuratClusteringCode">
                        Show/Hide Code
                    </button>
                    <!-- Collapsible code block -->
                    <div class="collapse" id="SeuratClusteringCode">
                        <pre><code class="language-r">
    st_data <- FindVariableFeatures(object=st_data)
    st_data <- ScaleData(object =st_data)
    st_data <- RunPCA(object = st_data)
    st_data <- RunUMAP(object = st_data, dims = 1:20)
    st_data <- FindNeighbors(object = st_data, dims = 1:20)
    st_data <- FindClusters(object = st_data)


</code></pre>

<<<<<<< HEAD
                        <img src="images/dimPlot2D.png" alt="Description of image" style="width:600px; height:auto;">
=======
                        <img src="images/LogNormalize.png" alt="Description of image" style="width:600px; height:auto;">
>>>>>>> 08d4d3d30d0a1794cbca0b57ae26f3137472227a


                    </div>



                    <!-- Expression Enhancement -->
                    <h2 id="Expression Enhancement" class="mt-5">Expression Enhancement</h2>
                    <h3>BayesSpace</h3>
                    <!-- Code block toggle button -->
                    <button class="btn btn-success float-right" type="button" data-toggle="collapse" data-target="#BayesSpace" aria-expanded="false" aria-controls="BayesSpace">
                        Show/Hide Code
                    </button>

                    <!-- Collapsible code block -->
                    <div class="collapse" id="BayesSpace">
                        <pre><code class="language-r">
          data=readRDS("SS200000135TL_D1.tissue.gef_bin20.seurat.rds")
          st=as.SingleCellExperiment(data)
          colData(st)$col=colData(st)$x/200
          colData(st)$row=colData(st)$y/200
          colData(st)$imagecol=colData(st)$x
          colData(st)$imagerow=colData(st)$y
          
          st <- spatialPreprocess(st, platform="ST" ,
                n.PCs=7, n.HVGs=2000, log.normalize=FALSE)
                st.tune <- qTune(st, qs=seq(2, 20), platform="ST", d=7)

          qPlot(st.tune)

          st.tune <- spatialCluster(st.tune, q=17, platform="ST", d=7,
                                     init.method="mclust", model="t",
                                     gamma=2,
                                     nrep=1000, burn.in=100,
                                     save.chain=TRUE)


          st.enhanced <- spatialEnhance(st.tune, q=17, platform="ST", d=7,
                                              model="t", gamma=2,
                                              jitter_prior=0.3, jitter_scale=3.5,
                                              nrep=1000, burn.in=100,
                                              save.chain=TRUE)

          markers=c("Ntng1")
          st.enhanced <- enhanceFeatures(st.enhanced, st,
                                               feature_names=markers,
                                               nrounds=0)
          featurePlot(st, "Ntng1")
          featurePlot(st.enhanced, "Ntng1")
        </code></pre>
                    </div>
                    <img src="images/Image_20241029120419.png" alt="Description of image" style="width:600px; height:auto;">


                    <!-- Spatial Domain Detection Section -->
                    <h2 id="Spatial Domain Detection" class="mt-5">Spatial Domain Detection</h2>
                    <p>
                        Spatial Domain Detection analysis in spatial transcriptomics involves identifying and characterizing distinct tissue regions based on gene expression patterns.
                        By analyzing the spatial organization of genes, this method segments the tissue into biologically meaningful domains, revealing how gene expression varies across different tissue structures.
                        The analysis often uses dimensionality reduction and clustering algorithms to group spots with similar expression profiles,
                        helping to uncover spatially resolved functional zones, which can provide insights into tissue architecture, cell-type distribution, and regional biological processes.
                    </p>

                    <h3>LogNormalize</h3>

                    <!-- Code block toggle button -->
                    <button class="btn btn-success float-right" type="button" data-toggle="collapse" data-target="#SpatialDDCode" aria-expanded="false" aria-controls="SpatialDDCode">
                        Show/Hide Code
                    </button>

                    <!-- Collapsible code block -->
                    <div class="collapse" id="SpatialDDCode">
                        <pre><code class="language-r">
library(Matrix)
library(Seurat)

meta_data <- st_data@meta.data
             # Create a data frame with ID, X, and Y columns
             formatted_coordinates <- data.frame(
  ID = rownames(meta_data),
  X = meta_data$x,
  Y = meta_data$y
)

# Save to a text file
write.table(formatted_coordinates, "./SS200000135TL_D1.tissue.gef_bin20.seurat.coordinates.txt", sep = "\t", row.names = FALSE, quote = FALSE)

formatted_metadata <- data.frame(
  CellID = rownames(meta_data),
  nCount = meta_data$nCount_Spatial,
  nFeature = meta_data$nCount_Spatial,
  mito=meta_data$percent.mito,
  x=meta_data$x,
  y=meta_data$y
  )

# Save to a text file
write.table(formatted_metadata, "./SS200000135TL_D1.tissue.gef_bin20.seurat.metadata.txt", sep = "\t", row.names = FALSE, quote = FALSE)

expression_matrix <- as.matrix(st_data@assays$RNA@counts)
expression_df <- data.frame(expression_matrix)
rownames(expression_df) <- rownames(expression_matrix)
write.table(expression_df, "./SS200000135TL_D1.tissue.gef_bin20.seurat.expression_matrix.txt", sep = "\t", quote = FALSE, col.names = NA)

</code></pre>


                    </div>



                </div>
        </div>
    </div>

    <script src="https://code.jquery.com/jquery-3.3.1.slim.min.js" integrity="sha384-q8i/X+965DzO0rT7abK41JStQIAqVgRVzpbzo5smXKp4YfRvH+8abtTE1Pi6jizo" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.7/umd/popper.min.js" integrity="sha384-UO2eT0CpHqdSJQ6hJty5KVphtPhzWj9WO1clHTMGa3JDZwrnQq4sF86dIHNDz0W1" crossorigin="anonymous"></script>
    <script src="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/js/bootstrap.min.js" integrity="sha384-ChfqqxuZUCnJSK3+MXmPNIyE6ZbWh2IMqE241rYiqJxyMiZ6OW/JmZQ5stwEULTy" crossorigin="anonymous"></script>

    <!-- Highlight.js initialization -->
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.5.1/styles/default.min.css">
    <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.5.1/highlight.min.js"></script>
    <script>hljs.highlightAll();</script>
</body>

</html>
